core/vote/vote_pool.go

```diff
@@ -0,0 +1,401 @@
+		heap.Push(votesPq, voteData)
+		voteBox := &VoteBox{
+			blockNumber:  targetNumber,
+			voteMessages: make([]*types.VoteEnvelope, 0, maxCurVoteAmountPerBlock),
+		}
+		m[targetHash] = voteBox
+
+		if isFutureVote {
+			localFutureVotesPqGauge.Inc(1)
+		} else {
+			localCurVotesPqGauge.Inc(1)
+		}
+	}
+
+	// Put into corresponding votes map.
+	m[targetHash].voteMessages = append(m[targetHash].voteMessages, vote)
+	// Add into received vote to avoid future duplicated vote comes.
+	pool.receivedVotes.Add(voteHash)
+	log.Info("VoteHash put into votepool is:", "voteHash", voteHash)
+
+	if isFutureVote {
+		localFutureVotesGauge.Inc(1)
+	} else {
+		localCurVotesGauge.Inc(1)
+	}
+	votesPerBlockHashMetric(targetHash).Inc(1)
+	localReceivedVotesGauge.Inc(1)
+}
+
+func (pool *VotePool) transferVotesFromFutureToCur(latestBlockHeader *types.Header) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	futurePq := pool.futureVotesPq
+	latestBlockNumber := latestBlockHeader.Number.Uint64()
+
+	// For vote before latestBlockHeader-13, transfer to cur if valid.
+	for futurePq.Len() > 0 && futurePq.Peek().TargetNumber+upperLimitOfVoteBlockNumber < latestBlockNumber {
+		blockHash := futurePq.Peek().TargetHash
+		pool.transfer(blockHash)
+	}
+
+	// For vote within latestBlockHeader-13 ~ latestBlockHeader, only transfer the the vote inside the local fork.
+	futurePqBuffer := make([]*types.VoteData, 0)
+	for futurePq.Len() > 0 && futurePq.Peek().TargetNumber <= latestBlockNumber {
+		blockHash := futurePq.Peek().TargetHash
+		header := pool.chain.GetHeaderByHash(blockHash)
+		if header == nil {
+			// Put into pq buffer used for later put again into futurePq
+			futurePqBuffer = append(futurePqBuffer, heap.Pop(futurePq).(*types.VoteData))
+			continue
+		}
+		pool.transfer(blockHash)
+	}
+
+	for _, voteData := range futurePqBuffer {
+		heap.Push(futurePq, voteData)
+	}
+}
+
+func (pool *VotePool) transfer(blockHash common.Hash) {
+	curPq, futurePq := pool.curVotesPq, pool.futureVotesPq
+	curVotes, futureVotes := pool.curVotes, pool.futureVotes
+
+	voteBox := futureVotes[blockHash]
+
+	validVotes := make([]*types.VoteEnvelope, 0, len(voteBox.voteMessages))
+	for _, vote := range voteBox.voteMessages {
+		// Verify if the vote comes from valid validators based on voteAddress (BLSPublicKey).
+		if posa, ok := pool.engine.(consensus.PoSA); ok {
+			if !posa.VerifyVote(pool.chain, vote) {
+				continue
+			}
+		}
+
+		// In the process of transfer, send valid vote to votes channel for handler usage
+		voteEv := core.NewVoteEvent{Vote: vote}
+		pool.votesFeed.Send(voteEv)
+		validVotes = append(validVotes, vote)
+	}
+
+	voteData := heap.Pop(futurePq)
+	if _, ok := curVotes[blockHash]; !ok {
+		heap.Push(curPq, voteData)
+		curVotes[blockHash] = &VoteBox{voteBox.blockNumber, validVotes}
+		localCurVotesPqGauge.Inc(1)
+	} else {
+		curVotes[blockHash].voteMessages = append(curVotes[blockHash].voteMessages, validVotes...)
```

Comment:
Is it possible that duplicated votes been appended? Filter the duplicated ones?

Reply:
We already have a receivedVotes mapset to filter duplicated votes, and it will be handled in basicVerify() func